ui/raidboss/emulator/data/AnalyzedEncounter.ts

```diff
@@ -0,0 +1,225 @@
+import { RaidbossOptions } from '../../raidboss_options';
+import Encounter from './Encounter';
+import RaidEmulator from './RaidEmulator';
+import LineEvent from './network_log_converter/LineEvent';
+import { UnreachableCode } from '../../../../resources/not_reached';
+import { LooseTrigger } from '../../../../types/trigger';
+import Combatant from './Combatant';
+import { PlayerChangedDetail } from '../../../../resources/player_override';
+
+type PerspectiveTrigger = {
+  triggerHelper: TriggerHelper;
+  status: ResolverStatus;
+  logLine: LineEvent;
+  resolvedOffset: number;
+};
+type Perspective = {
+  initialData: DataType;
+  triggers: PerspectiveTrigger[];
+  finalData?: DataType;
+};
+type Perspectives = { [id: string]: Perspective };
+
+export default class AnalyzedEncounter extends EventBus {
+  popupText?: PopupTextAnalysis;
+  perspectives: Perspectives = {};
+  constructor(
+    public options: RaidbossOptions,
+    public encounter: Encounter,
+    public emulator: RaidEmulator) {
+    super();
+  }
+
+  selectPerspective(ID: string): void {
+    if (this.encounter && this.encounter.combatantTracker) {
+      const selectedPartyMember = this.encounter.combatantTracker.combatants[ID];
+      if (!selectedPartyMember)
+        return;
+
+      this.popupText?.getPartyTracker().onPartyChanged({
+        party: this.encounter.combatantTracker.partyMembers.map((ID) => {
+          const partyMember = this.encounter?.combatantTracker?.combatants[ID];
+          if (!partyMember)
+            throw new UnreachableCode();
+          return {
+            id: ID,
+            worldId: 0,
+            name: partyMember.name,
+            job: Util.jobToJobEnum(partyMember.job ?? 'NONE'),
+            inParty: true,
+          };
+        }),
+      });
+      this.updateState(selectedPartyMember, this.encounter.startTimestamp);
+      this.popupText?.OnChangeZone({
+        type: 'ChangeZone',
+        zoneName: this.encounter.encounterZoneName,
+        zoneID: parseInt(this.encounter.encounterZoneId, 16),
+      });
+    }
+  }
+
+  updateState(combatant: Combatant, timestamp: number): void {
+    const job = combatant.job;
+    if (!job)
+      throw new UnreachableCode();
+    const state = combatant.getState(timestamp);
+    const eventCast = {
+      detail: {
+        name: combatant.name,
+        job: job,
+        level: combatant.level ?? 0,
+        currentHP: state.hp,
+        maxHP: state.maxHp,
+        currentMP: state.mp,
+        maxMP: state.maxMp,
+        currentCP: 0,
+        maxCP: 0,
+        currentGP: 0,
+        maxGP: 0,
+        currentShield: 0,
+        jobDetail: null,
+        pos: {
+          x: state.posX,
+          y: state.posY,
+          z: state.posZ,
+        },
+        rotation: state.heading,
+        bait: 0,
+        debugJob: '',
+      },
+    };
+    // For emulator purposes, jobDetail is always null, but typescript really doesn't like that
```

Comment:
I FIGURED IT OUT.  \*\o/\*

The issue is not `jobDetail` as you might suspect, but instead the issue is `job`.  Because `Job` is a union type, `T extends Job` distributes the `Job` union across all possibilities leading to an incompatible / impossible type to initialize.

This is the error:
```
      TS2322: Type '{ name: string; job: Job; level: number; currentHP: number;
maxHP: number; currentMP: number; maxMP: number; currentCP: number; maxCP: numbe
r; currentGP: number; maxGP: number; currentShield: number; ... 4 more ...; debu
gJob: string; }' is not assignable to type '{ name: string; job: "NONE"; level:
number; currentHP: number; maxHP: number; currentMP: number; maxMP: number; curr
entCP: number; maxCP: number; currentGP: number; maxGP: number; currentShield: n
umber; ... 4 more ...; debugJob: string; } | ... 37 more ... | { ...; }'.
  Type '{ name: string; job: Job; level: number; currentHP: number; maxHP: numbe
r; currentMP: number; maxMP: number; currentCP: number; maxCP: number; currentGP
: number; maxGP: number; currentShield: number; ... 4 more ...; debugJob: string
; }' is not assignable to type '{ name: string; job: "DNC"; level: number; curre
ntHP: number; maxHP: number; currentMP: number; maxMP: number; currentCP: number
; maxCP: number; currentGP: number; maxGP: number; currentShield: number; ... 4
more ...; debugJob: string; }'.
    Types of property 'job' are incompatible.
      Type 'Job' is not assignable to type '"DNC"'.
        Type '"NONE"' is not assignable to type '"DNC"'.
```

You can verify that job is the issue because you can say some nonsense like `job: job as 'DNC'` and it will then become the correct type.

You can fix the underlying issue with `type PlayerChangedRet = Job extends infer T ? [T] extends [Job] ? {`.

I ran into this same thing earlier this week and https://www.reddit.com/r/typescript/comments/gmjfxe/weirdness_with_booleans/ solved it then as well.

Reply:
That doesn't actually fix the problem. `jobDetail` just ends up always typed as `unknown` even when testing `job`.

![image](https://user-images.githubusercontent.com/6119598/122253527-3edeb400-ce9a-11eb-9fd9-363579bc8d61.png)
![image](https://user-images.githubusercontent.com/6119598/122253576-49994900-ce9a-11eb-9f5d-d396488de4de.png)

Reply:
Whoooooooops.  That's what I get for posting something before bed.  I think I get it now.

(Explaining this to myself out loud for future me who is reading this, since you likely already know it.)  You need the union distributed because that's what causes it to be a discriminated union that's able to tell the types apart.  Using the brackets makes it no longer a discriminated union, so `jobDetail` becomes unknown again.  Unfortunately you can't easily construct discriminated unions from variables, because they need to be statically typed in some way (e.g. [this problem](https://stackoverflow.com/questions/51064527/defining-a-variable-as-one-variant-of-a-discriminated-union-in-typescript).

The only option I could come up with after some (minor) sleep is to add another case where we use `null` as another case in the discriminated union.  In other words, if you set `jobDetail: null` then `job` can be anything.  This allows you to at least construct your one case (but doesn't solve synthetically generating anything with a real `jobDetail`), e.g.

```typescript
type PlayerChangedJobDetails<T> = {
  job: T;
  jobDetail: JobDetail[T];
} | {
  job: Job;
  jobDetail: null;
}

type PlayerChangedBase = {
  name: string;
  level: number;
  currentHP: number;
  maxHP: number;
  currentMP: number;
  maxMP: number;
  currentCP: number;
  maxCP: number;
  currentGP: number;
  maxGP: number;
  currentShield: number;
  pos: {
    x: number;
    y: number;
    z: number;
  };
  rotation: number;
  bait: number;
  debugJob: string;
};

type PlayerChangedRet = Job extends infer T ? T extends Job ?
  PlayerChangedJobDetails<T> & PlayerChangedBase : never : never;
```

Reply:
This actually seems to work for constructing one wholesale as well. e.g. the following work.

Made this adjustment and removed the casting.

```typescript
    const detail: PlayerChangedDetail = {
      detail: {
        name: combatant.name,
        job: 'WHM',
        level: combatant.level ?? 0,
        currentHP: state.hp,
        maxHP: state.maxHp,
        currentMP: state.mp,
        maxMP: state.maxMp,
        currentCP: 0,
        maxCP: 0,
        currentGP: 0,
        maxGP: 0,
        currentShield: 0,
        jobDetail: {
          lilyStacks: 0,
          lilyMilliseconds: 0,
          bloodlilyStacks: 0,
        },
        pos: {
          x: state.posX,
          y: state.posY,
          z: state.posZ,
        },
        rotation: state.heading,
        bait: 0,
        debugJob: '',
      },
    };
```

```typescript
    const job: string = 'WHM';
    const detail = {
      name: combatant.name,
      level: combatant.level ?? 0,
      currentHP: state.hp,
      maxHP: state.maxHp,
      currentMP: state.mp,
      maxMP: state.maxMp,
      currentCP: 0,
      maxCP: 0,
      currentGP: 0,
      maxGP: 0,
      currentShield: 0,
      pos: {
        x: state.posX,
        y: state.posY,
        z: state.posZ,
      },
      rotation: state.heading,
      bait: 0,
      debugJob: '',
    };
    if (job === 'WHM') {
      const fullDetail: PlayerChangedDetail = {
        detail: {
          ...detail,
          job: 'WHM',
          jobDetail: {
            lilyStacks: 0,
            lilyMilliseconds: 0,
            bloodlilyStacks: 0,
          },
        },
      };
    }
```

Reply:
> This actually seems to work for constructing one wholesale as well. e.g. the following work.

I could be wrong, but I think those work (and would have worked before my suggestion) because in both cases TypeScript knows statically that job is `WHM`.  I think the heart of the issue is that you can't assign a job whose type is just `Job`.  If the `Job` is literally the string `WHM` or you do a conditional where you verify that it is `WHM` then that's enough information to know statically.

Reply:
I'm going to mark this as resolved, since it's fixed from a code perspective for this PR. Might be something to revisit later on though.